File: 355. Design Twitter.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
public:
    int userId;
    int tweetId;
    Node* next;
    Node(int uId, int tId) : userId(uId), tweetId(tId), next(nullptr) {}
};

class LinkedList {
public:
    Node* head;

    LinkedList() : head(nullptr) {};

    void insertAtBegin(int userId, int tweetId) {
        Node * newNode = new Node(userId, tweetId);
        newNode->next = this->head;
        this->head = newNode;
    }
};

class Twitter {
public:

    unordered_map<int, set<int>> followerToFollowing;
    LinkedList userTweetsLL;
    int time = 1;

    Twitter() {
        
    }
    
    void postTweet(int userId, int tweetId) {
        userTweetsLL.insertAtBegin(userId, tweetId);
        if (followerToFollowing.find(userId) == followerToFollowing.end()){
            followerToFollowing[userId] = set<int>();
            followerToFollowing[userId].insert(userId);
        }
    }
    
    vector<int> getNewsFeed(int userId) {
         if (followerToFollowing.find(userId) == followerToFollowing.end()){
            followerToFollowing[userId] = set<int>();
            followerToFollowing[userId].insert(userId);
        }
        vector<int> tweetList;
        Node* iterator = userTweetsLL.head;
        int counter = 10;
        while (iterator && counter) {
            if (
                followerToFollowing[userId].find(iterator->userId) 
                != followerToFollowing[userId].end() 
            ) {
                tweetList.push_back(iterator->tweetId);
                counter--;
            }
            iterator = iterator->next;
        }
        return tweetList;
    }
    
    void follow(int followerId, int followeeId) {
        if (followerToFollowing.find(followerId) == followerToFollowing.end()){
            followerToFollowing[followerId] = set<int>();
            followerToFollowing[followerId].insert(followerId);
        }
        followerToFollowing[followerId].insert(followeeId);
    }
    
    void unfollow(int followerId, int followeeId) {
        if (followerToFollowing[followerId].find(followeeId) 
        != followerToFollowing[followerId].end()) {
            followerToFollowing[followerId].erase(followeeId);
        }
    }
};

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter* obj = new Twitter();
 * obj->postTweet(userId,tweetId);
 * vector<int> param_2 = obj->getNewsFeed(userId);
 * obj->follow(followerId,followeeId);
 * obj->unfollow(followerId,followeeId);
 */

File: 355. Design Twitter.py
from typing import List

class LinkedList:
    def __init__(self):
        self.head = None
    
    def insert_at_begin(self, userId, tweetId):
        new_node = Node(userId=userId, tweetId=tweetId)
        new_node.next = self.head
        self.head = new_node
    

class Node:
    def __init__(self, userId, tweetId):
        self.userId = userId
        self.tweetId = tweetId
        self.next = None


class Twitter:

    def __init__(self):
        self.follower_to_following = dict()
        self.user_tweets_ll = LinkedList()
        self.time = 1

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.user_tweets_ll.insert_at_begin(userId=userId, tweetId=tweetId)
        if userId not in self.follower_to_following:
            self.follower_to_following[userId] = set()
            self.follower_to_following[userId].add(userId)
        

    def getNewsFeed(self, userId: int) -> List[int]:
        if userId not in self.follower_to_following:
            self.follower_to_following[userId] = set()
            self.follower_to_following[userId].add(userId)
        tweet_list = list()
        iterator = self.user_tweets_ll.head
        counter = 10
        while iterator and counter:
            if iterator.userId in self.follower_to_following[userId]:
                tweet_list.append(iterator.tweetId)
                counter -= 1
            iterator = iterator.next
        
        return tweet_list


    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId not in self.follower_to_following:
            self.follower_to_following[followerId] = set()
            self.follower_to_following[followerId].add(followerId)
        self.follower_to_following[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followeeId in self.follower_to_following[followerId]:
            self.follower_to_following[followerId].remove(followeeId)


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)